File: apps/mobile/app/lib/features/trip_history/data/trip_history_api.dart
import 'package:dio/dio.dart';

import '../domain/trip_history_full_read.dart';
import '../domain/trip_history_short_info_read.dart';

class TripHistoryListQuery {
  const TripHistoryListQuery({
    this.userId,
    this.limit,
    this.offset,
    this.startedAfter,
    this.startedBefore,
    this.finishedAfter,
    this.finishedBefore,
  });

  final String? userId;
  final int? limit;
  final int? offset;
  final String? startedAfter;
  final String? startedBefore;
  final String? finishedAfter;
  final String? finishedBefore;

  Map<String, dynamic> toQueryParameters() {
    final m = <String, dynamic>{};
    void put(String k, Object? v) {
      if (v == null) return;
      if (v is String && v.isEmpty) return;
      m[k] = v;
    }

    put('userId', userId);
    if (limit != null) put('limit', limit);
    if (offset != null) put('offset', offset);
    put('startedAfter', startedAfter);
    put('startedBefore', startedBefore);
    put('finishedAfter', finishedAfter);
    put('finishedBefore', finishedBefore);
    return m;
  }
}

class TripHistoryApi {
  TripHistoryApi(this._dio);

  final Dio _dio;

  Future<List<TripHistoryShortInfoRead>> listShort(
    TripHistoryListQuery query,
  ) async {
    final res = await _dio.get<List<dynamic>>(
      '/trip-history',
      queryParameters: query.toQueryParameters(),
    );
    final data = res.data ?? <dynamic>[];
    return data
        .whereType<Map>()
        .map((e) => TripHistoryShortInfoRead.fromJson(
              e.map((k, v) => MapEntry(k.toString(), v)),
            ))
        .toList(growable: false);
  }

  Future<TripHistoryShortInfoRead> getShort(String tripId) async {
    final res = await _dio.get<Map<String, dynamic>>(
      '/trip-history/${Uri.encodeComponent(tripId)}',
    );
    final m = res.data;
    if (m == null) {
      throw StateError('trip-history: empty body');
    }
    return TripHistoryShortInfoRead.fromJson(m);
  }

  Future<TripHistoryFullRead> getFull(String tripId) async {
    final res = await _dio.get<Map<String, dynamic>>(
      '/trip-history/${Uri.encodeComponent(tripId)}/full',
    );
    final m = res.data;
    if (m == null) {
      throw StateError('trip-history/full: empty body');
    }
    return TripHistoryFullRead.fromJson(m);
  }
}

File: apps/web/src/features/trips/api/trip-history.api.ts
import type {
  TripHistoryFullRead,
  TripHistoryShortInfoRead,
} from "@/entities/trip/model/trip-history-contracts";
import { BaseApiClient } from "@/shared/api";
import type { AccessTokenGetter } from "@/shared/api/base-api-client";

function optionalQuery(params: Record<string, string | undefined>): string {
  const sp = new URLSearchParams();
  for (const [k, v] of Object.entries(params)) {
    if (v !== undefined && v !== "") {
      sp.set(k, v);
    }
  }
  const q = sp.toString();
  return q ? `?${q}` : "";
}

export type TripHistoryListQuery = {
  /** Для MANAGER/SYSTEM_ADMIN — чья история (водитель и т.д.). */
  userId?: string;
  limit?: number;
  offset?: number;
  /** ISO-8601, границы по `trip.startedAt`. */
  startedAfter?: string;
  startedBefore?: string;
  /** Только завершённые по `finishedAt`. */
  finishedAfter?: string;
  finishedBefore?: string;
};

export class TripHistoryApi extends BaseApiClient {
  constructor(baseUrl: string, getAccessToken: AccessTokenGetter) {
    super(baseUrl, getAccessToken);
  }

  /** Список поездок с нарушениями по пользователю (`GET /trip-history`). */
  listShort(
    query: TripHistoryListQuery = {},
  ): Promise<TripHistoryShortInfoRead[]> {
    return this.getJson<TripHistoryShortInfoRead[]>(
      `/trip-history${optionalQuery({
        userId: query.userId,
        limit: query.limit !== undefined ? String(query.limit) : undefined,
        offset: query.offset !== undefined ? String(query.offset) : undefined,
        startedAfter: query.startedAfter,
        startedBefore: query.startedBefore,
        finishedAfter: query.finishedAfter,
        finishedBefore: query.finishedBefore,
      })}`,
    );
  }

  /** Карточка поездки для истории: поездка + авто + нарушения (без точек телеметрии). */
  getShort(tripId: string): Promise<TripHistoryShortInfoRead> {
    return this.getJson<TripHistoryShortInfoRead>(
      `/trip-history/${encodeURIComponent(tripId)}`,
    );
  }

  /** Полная карточка: поездка + авто + нарушения + точки телеметрии. */
  getFull(tripId: string): Promise<TripHistoryFullRead> {
    return this.getJson<TripHistoryFullRead>(
      `/trip-history/${encodeURIComponent(tripId)}/full`,
    );
  }
}
